File: scripts/sync-nepse-directory.ts
import { createClient } from "@supabase/supabase-js";
import {
  getSecurityList,
  getCompaniesList,
  disclosure,
  shutdownWorkerPool,
} from "nepse-api-unofficial";

const url = process.env.SUPABASE_URL;
const key = process.env.SUPABASE_SERVICE_ROLE_KEY;

if (!url || !key) {
  throw new Error("Missing SUPABASE_URL or SUPABASE_SERVICE_ROLE_KEY.");
}

const sb = createClient(url, key, {
  auth: { persistSession: false },
});

function asArray(value: any): any[] {
  if (Array.isArray(value)) return value;

  for (const key of [
    "data",
    "content",
    "securities",
    "companies",
    "companyList",
    "securityList",
    "disclosures",
  ]) {
    if (Array.isArray(value?.[key])) return value[key];
  }

  return [];
}

function first(...values: any[]) {
  return values.find(
    (value) =>
      value !== undefined &&
      value !== null &&
      String(value).trim() !== ""
  );
}

function clean(value: any): string | null {
  const result = value == null ? "" : String(value).trim();
  return result || null;
}

function normalizeDate(value: any): string | null {
  if (!value) return null;

  const parsed = new Date(value);
  return Number.isNaN(parsed.getTime())
    ? null
    : parsed.toISOString();
}

async function main() {
  console.log("Downloading NEPSE security and company directories...");

  const [securityResult, companyResult, disclosureResult] =
    await Promise.all([
      getSecurityList(),
      getCompaniesList(),
      disclosure(),
    ]);

  const securities = asArray(securityResult);
  const companies = asArray(companyResult);
  const disclosures = asArray(disclosureResult);

  console.log(`Security records received: ${securities.length}`);
  console.log(`Company records received: ${companies.length}`);
  console.log(`Disclosure records received: ${disclosures.length}`);

  const companyBySymbol = new Map<string, any>();

  for (const row of companies) {
    const symbol = clean(
      first(
        row.symbol,
        row.securitySymbol,
        row.companySymbol,
        row.stockSymbol
      )
    )?.toUpperCase();

    if (symbol) companyBySymbol.set(symbol, row);
  }

  const securityRows = securities
    .map((row) => {
      const symbol = clean(
        first(
          row.symbol,
          row.securitySymbol,
          row.stockSymbol,
          row.companySymbol
        )
      )?.toUpperCase();

      if (!symbol) return null;

      const company = companyBySymbol.get(symbol);

      const securityName = clean(
        first(
          row.securityName,
          row.security_name,
          row.name,
          row.companyName,
          company?.companyName,
          company?.company_name,
          company?.name
        )
      );

      const companyName = clean(
        first(
          row.companyName,
          row.company_name,
          row.issuerName,
          row.issuer_name,
          company?.companyName,
          company?.company_name,
          company?.name,
          securityName
        )
      );

      return {
        symbol,
        security_name: securityName,
        company_name: companyName,
      };
    })
    .filter(Boolean);

  const merged = new Map<string, any>();

  for (const row of securityRows) {
    merged.set(row.symbol, row);
  }

  for (const [symbol, company] of companyBySymbol) {
    const current = merged.get(symbol) || { symbol };

    merged.set(symbol, {
      symbol,
      security_name:
        current.security_name ||
        clean(
          first(
            company.securityName,
            company.security_name,
            company.companyName,
            company.company_name,
            company.name
          )
        ),
      company_name:
        current.company_name ||
        clean(
          first(
            company.companyName,
            company.company_name,
            company.name,
            company.securityName
          )
        ),
    });
  }

  const directoryRows = [...merged.values()].filter(
    (row) => row.symbol
  );

  if (!directoryRows.length) {
    throw new Error(
      "NEPSE returned no usable security-directory records."
    );
  }

  const { error: securityError } = await sb
    .from("nepse_securities")
    .upsert(directoryRows, {
      onConflict: "symbol",
    });

  if (securityError) throw securityError;

  const aliasRows = directoryRows.map((row) => ({
    symbol: row.symbol,
    company_name:
      row.company_name || row.security_name || null,
    aliases: [
      row.security_name,
      row.company_name,
    ].filter(
      (value, index, array) =>
        value &&
        array.indexOf(value) === index
    ),
  }));

  const { error: aliasError } = await sb
    .from("company_aliases")
    .upsert(aliasRows, {
      onConflict: "symbol",
    });

  if (aliasError) throw aliasError;

  const disclosureRows = disclosures
    .map((row) => {
      const symbol = clean(
        first(
          row.symbol,
          row.securitySymbol,
          row.stockSymbol,
          row.companySymbol
        )
      )?.toUpperCase();

      const title = clean(
        first(
          row.title,
          row.newsHeadline,
          row.headline,
          row.subject,
          row.description
        )
      );

      const publishedAt = normalizeDate(
        first(
          row.publishedAt,
          row.publishedDate,
          row.newsDate,
          row.createdAt,
          row.date
        )
      );

      if (!symbol || !title || !publishedAt) return null;

      return {
        symbol,
        title,
        published_at: publishedAt,
        category:
          clean(
            first(
              row.category,
              row.newsType,
              row.disclosureType
            )
          ) || "NEPSE disclosure",
        source_url: clean(
          first(
            row.url,
            row.newsUrl,
            row.sourceUrl,
            row.attachmentUrl
          )
        ),
        source_name: "Nepal Stock Exchange",
        summary: clean(
          first(
            row.description,
            row.details,
            row.summary
          )
        ),
        authority_score: 95,
        fetched_at: new Date().toISOString(),
      };
    })
    .filter(Boolean);

  if (disclosureRows.length) {
    const { error: disclosureError } = await sb
      .from("nepse_disclosures")
      .upsert(disclosureRows, {
        onConflict: "symbol,title,published_at",
      });

    if (disclosureError) throw disclosureError;
  }

  console.log(
    `Saved ${directoryRows.length} security-directory records.`
  );
  console.log(
    `Saved ${aliasRows.length} company alias records.`
  );
  console.log(
    `Saved ${disclosureRows.length} NEPSE disclosures.`
  );
}

try {
  await main();
} finally {
  await shutdownWorkerPool();
}

File: scripts/sync-nepse-directory.py
import json
import os
import sys
from typing import Any

import requests
from nepse import Nepse


SUPABASE_URL = os.environ.get("SUPABASE_URL", "").rstrip("/")
SERVICE_KEY = os.environ.get("SUPABASE_SERVICE_ROLE_KEY", "")

if not SUPABASE_URL or not SERVICE_KEY:
    raise RuntimeError(
        "Missing SUPABASE_URL or SUPABASE_SERVICE_ROLE_KEY."
    )

HEADERS = {
    "apikey": SERVICE_KEY,
    "Authorization": f"Bearer {SERVICE_KEY}",
    "Content-Type": "application/json",
    "Prefer": "resolution=merge-duplicates,return=minimal",
}


def first_value(row: dict[str, Any], names: list[str]) -> Any:
    for name in names:
        value = row.get(name)
        if value is not None and str(value).strip():
            return value
    return None


def normalize_text(value: Any) -> str | None:
    if value is None:
        return None
    text = str(value).strip()
    return text or None


def extract_rows(value: Any) -> list[dict[str, Any]]:
    if isinstance(value, list):
        return [row for row in value if isinstance(row, dict)]

    if isinstance(value, dict):
        for key in (
            "data",
            "content",
            "companies",
            "companyList",
            "company_list",
            "securities",
        ):
            rows = value.get(key)
            if isinstance(rows, list):
                return [row for row in rows if isinstance(row, dict)]

    return []


def post_rows(table: str, rows: list[dict[str, Any]]) -> None:
    if not rows:
        return

    endpoint = f"{SUPABASE_URL}/rest/v1/{table}"

    for start in range(0, len(rows), 250):
        chunk = rows[start : start + 250]
        response = requests.post(
            endpoint,
            headers=HEADERS,
            data=json.dumps(chunk),
            timeout=60,
        )

        if not response.ok:
            raise RuntimeError(
                f"Supabase {table} upsert failed "
                f"({response.status_code}): {response.text}"
            )

        print(
            f"Saved {min(start + len(chunk), len(rows))}"
            f"/{len(rows)} rows to {table}"
        )


def main() -> None:
    print("Downloading the NEPSE company directory...")

    nepse = Nepse()

    # The library's own documentation recommends disabling TLS
    # verification while NEPSE's certificate-chain issue persists.
    nepse.setTLSVerification(False)

    raw_result = nepse.getCompanyList()
    companies = extract_rows(raw_result)

    print(f"Raw company records received: {len(companies)}")

    if not companies:
        print("Returned value:", repr(raw_result)[:2000])
        raise RuntimeError("NEPSE returned no company-directory rows.")

    directory_by_symbol: dict[str, dict[str, Any]] = {}

    for row in companies:
        symbol = normalize_text(
            first_value(
                row,
                [
                    "symbol",
                    "securitySymbol",
                    "security_symbol",
                    "companySymbol",
                    "stockSymbol",
                    "ticker",
                ],
            )
        )

        company_name = normalize_text(
            first_value(
                row,
                [
                    "companyName",
                    "company_name",
                    "securityName",
                    "security_name",
                    "companyNameEnglish",
                    "name",
                ],
            )
        )

        if not symbol:
            continue

        symbol = symbol.upper()

        directory_by_symbol[symbol] = {
            "symbol": symbol,
            "security_name": company_name,
            "company_name": company_name,
        }

    directory_rows = list(directory_by_symbol.values())

    if not directory_rows:
        print(
            "Example raw record:",
            json.dumps(companies[0], ensure_ascii=False, indent=2),
        )
        raise RuntimeError(
            "Company rows were returned, but no symbols could be parsed."
        )

    alias_rows = [
        {
            "symbol": row["symbol"],
            "company_name": row["company_name"],
            "aliases": (
                [row["company_name"]]
                if row["company_name"]
                else []
            ),
        }
        for row in directory_rows
    ]

    post_rows("nepse_securities", directory_rows)
    post_rows("company_aliases", alias_rows)

    names = sum(
        1
        for row in directory_rows
        if row["company_name"]
    )

    print(f"Directory records saved: {len(directory_rows)}")
    print(f"Symbols with company names: {names}")

    if names == 0:
        raise RuntimeError(
            "Directory saved, but no company names were returned."
        )


if __name__ == "__main__":
    try:
        main()
    except Exception as exc:
        print(f"ERROR: {exc}", file=sys.stderr)
        raise
